File: src/services/docxPreProcessor.js
/**
 * DOCX Pre-Processor
 *
 * Normalizes and fixes DOCX formatting before LibreOffice conversion
 * to improve fidelity and reduce styling issues.
 *
 * Fixes:
 * - Theme colors -> RGB colors
 * - Custom fonts -> standard fonts
 * - Complex styles -> simplified styles
 * - Word-specific features -> LibreOffice-compatible equivalents
 * - Paragraph spacing normalization
 * - Font weight/bold consistency
 */

const fs = require('fs');
const path = require('path');
const AdmZip = require('adm-zip');
const { XMLParser, XMLBuilder } = require('fast-xml-parser');

class DocxPreProcessor {
  constructor() {
    // Font substitution map - replace custom/theme fonts with LibreOffice-friendly fonts
    this.fontMap = {
      // Office theme fonts -> standard fonts
      'Calibri Light': 'Calibri',
      'Segoe UI Light': 'Segoe UI',

      // Custom fonts -> safe alternatives
      'Aptos': 'Calibri',
      'Aptos Narrow': 'Arial Narrow',
      'Arial Narrow': 'Arial',

      // Proprietary fonts -> open equivalents
      'Times New Roman': 'Liberation Serif',
      'Arial': 'Liberation Sans',
      'Courier New': 'Liberation Mono',

      // Keep these safe fonts as-is (LibreOffice handles well)
      'Calibri': 'Calibri',
      'Verdana': 'Verdana',
      'Georgia': 'Georgia',
      'Tahoma': 'Tahoma'
    };

    // Theme color -> RGB mappings (common Office theme colors)
    this.themeColorMap = {
      'accent1': '4472C4',  // Blue
      'accent2': 'ED7D31',  // Orange
      'accent3': 'A5A5A5',  // Gray
      'accent4': 'FFC000',  // Yellow
      'accent5': '5B9BD5',  // Light Blue
      'accent6': '70AD47',  // Green
      'dark1': '000000',    // Black
      'dark2': '44546A',    // Dark Gray
      'light1': 'FFFFFF',   // White
      'light2': 'E7E6E6'    // Light Gray
    };

    this.parserOptions = {
      ignoreAttributes: false,
      attributeNamePrefix: '@_',
      textNodeName: '#text',
      parseAttributeValue: false,
      trimValues: false
    };

    this.builderOptions = {
      ignoreAttributes: false,
      attributeNamePrefix: '@_',
      textNodeName: '#text',
      format: false,  // Don't format XML (preserve structure)
      suppressEmptyNode: false
    };
  }

  /**
   * Pre-process a DOCX file to improve LibreOffice compatibility
   *
   * @param {string} inputPath - Original DOCX file
   * @param {string} outputPath - Pre-processed DOCX file
   * @returns {Promise<Object>} Processing result with applied fixes
   */
  async process(inputPath, outputPath) {
    const fixes = {
      fontsNormalized: 0,
      themeColorsConverted: 0,
      stylesSimplified: 0,
      paragraphsAdjusted: 0,
      boldFixed: 0
    };

    try {
      console.log('Pre-processing DOCX for improved compatibility...');

      // Extract DOCX (it's a ZIP file)
      const zip = new AdmZip(inputPath);
      const zipEntries = zip.getEntries();

      // Parse XML files
      const parser = new XMLParser(this.parserOptions);
      const builder = new XMLBuilder(this.builderOptions);

      // Process document.xml (main content) - using safer regex approach
      const documentEntry = zip.getEntry('word/document.xml');
      if (documentEntry) {
        let documentXml = documentEntry.getData().toString('utf8');
        documentXml = this._processXmlWithRegex(documentXml, fixes);
        zip.updateFile('word/document.xml', Buffer.from(documentXml, 'utf8'));
      }

      // Process styles.xml (document styles) - using safer regex approach
      const stylesEntry = zip.getEntry('word/styles.xml');
      if (stylesEntry) {
        let stylesXml = stylesEntry.getData().toString('utf8');
        stylesXml = this._processXmlWithRegex(stylesXml, fixes);
        zip.updateFile('word/styles.xml', Buffer.from(stylesXml, 'utf8'));
      }

      // Do NOT remove theme - it causes issues
      // Instead, theme colors will be converted to RGB in-place

      // Save processed DOCX
      zip.writeZip(outputPath);

      console.log('Pre-processing complete:');
      console.log(`  • Fonts normalized: ${fixes.fontsNormalized}`);
      console.log(`  • Theme colors converted: ${fixes.themeColorsConverted}`);
      console.log(`  • Styles simplified: ${fixes.stylesSimplified}`);
      console.log(`  • Paragraphs adjusted: ${fixes.paragraphsAdjusted}`);
      console.log(`  • Bold formatting fixed: ${fixes.boldFixed}`);

      return {
        success: true,
        inputPath,
        outputPath,
        fixes
      };

    } catch (error) {
      console.error('DOCX pre-processing error:', error);
      throw new Error(`Pre-processing failed: ${error.message}`);
    }
  }

  /**
   * Process XML using safer regex-based replacements
   * Avoids restructuring XML which can break DOCX compatibility
   *
   * @param {string} xml - XML content
   * @param {Object} fixes - Fix counter
   * @returns {string} Modified XML
   * @private
   */
  _processXmlWithRegex(xml, fixes) {
    let modifiedXml = xml;

    // 1. Convert theme colors to RGB
    // Match: w:color w:themeColor="accent1"
    // Replace with explicit RGB color
    Object.entries(this.themeColorMap).forEach(([themeName, rgbColor]) => {
      const themePattern = new RegExp(`(<w:color[^>]*\\s)w:themeColor="${themeName}"([^>]*w:val=")[^"]*"`, 'g');
      const replacement = `$1w:val="${rgbColor}"`;
      const before = modifiedXml.length;
      modifiedXml = modifiedXml.replace(themePattern, replacement);
      if (modifiedXml.length !== before) {
        fixes.themeColorsConverted++;
      }
    });

    // 2. Normalize fonts - replace problem fonts
    Object.entries(this.fontMap).forEach(([oldFont, newFont]) => {
      if (oldFont !== newFont) {
        // Match font attributes
        const fontPattern = new RegExp(`(w:ascii="|w:hAnsi="|w:cs="|w:eastAsia=")${oldFont}"`, 'g');
        const before = modifiedXml.length;
        modifiedXml = modifiedXml.replace(fontPattern, `$1${newFont}"`);
        if (modifiedXml.length !== before) {
          fixes.fontsNormalized++;
        }
      }
    });

    // 3. Remove theme font references (force explicit fonts)
    const themeFontAttrs = ['w:asciiTheme', 'w:hAnsiTheme', 'w:cstheme', 'w:eastAsiaTheme'];
    themeFontAttrs.forEach(attr => {
      const pattern = new RegExp(`\\s${attr}="[^"]*"`, 'g');
      modifiedXml = modifiedXml.replace(pattern, '');
    });

    // 4. Remove unsupported text effects
    const unsupportedEffects = [
      'w:shadow',
      'w:outline',
      'w:emboss',
      'w:imprint',
      'w14:glow',
      'w14:shadow',
      'w14:reflection',
      'w14:textOutline',
      'w14:textFill'
    ];

    unsupportedEffects.forEach(effect => {
      // Remove entire element: <w:effect ... />
      const selfClosingPattern = new RegExp(`<${effect}[^/]*/>`, 'g');
      const before = modifiedXml.length;
      modifiedXml = modifiedXml.replace(selfClosingPattern, '');
      if (modifiedXml.length !== before) {
        fixes.stylesSimplified++;
      }

      // Remove element with closing tag: <w:effect ...>...</w:effect>
      const pairPattern = new RegExp(`<${effect}[^>]*>.*?</${effect}>`, 'g');
      modifiedXml = modifiedXml.replace(pairPattern, '');
    });

    // 5. Normalize AUTO color to explicit black
    modifiedXml = modifiedXml.replace(/w:val="AUTO"/g, 'w:val="000000"');

    // 6. Fix bold values (ensure explicit)
    modifiedXml = modifiedXml.replace(/<w:b\s*\/>/g, '<w:b w:val="1"/>');
    modifiedXml = modifiedXml.replace(/<w:b\s+w:val="true"\s*\/>/g, '<w:b w:val="1"/>');

    return modifiedXml;
  }

  /**
   * DEPRECATED - DOM-based processing (kept for reference)
   * Process main document content
   *
   * @param {Object} documentData - Parsed document XML
   * @param {Object} fixes - Fix counter object
   * @private
   */
  _processDocument_DEPRECATED(documentData, fixes) {
    if (!documentData['w:document'] || !documentData['w:document']['w:body']) {
      return;
    }

    const body = documentData['w:document']['w:body'];

    // Process paragraphs
    if (body['w:p']) {
      const paragraphs = Array.isArray(body['w:p']) ? body['w:p'] : [body['w:p']];

      paragraphs.forEach(paragraph => {
        this._processParagraph(paragraph, fixes);
      });
    }

    // Process tables
    if (body['w:tbl']) {
      const tables = Array.isArray(body['w:tbl']) ? body['w:tbl'] : [body['w:tbl']];

      tables.forEach(table => {
        this._processTable(table, fixes);
      });
    }
  }

  /**
   * Process a paragraph
   *
   * @param {Object} paragraph - Paragraph object
   * @param {Object} fixes - Fix counter
   * @private
   */
  _processParagraph(paragraph, fixes) {
    // Fix paragraph properties
    if (paragraph['w:pPr']) {
      this._normalizeParagraphProperties(paragraph['w:pPr'], fixes);
    }

    // Process runs (text segments)
    if (paragraph['w:r']) {
      const runs = Array.isArray(paragraph['w:r']) ? paragraph['w:r'] : [paragraph['w:r']];

      runs.forEach(run => {
        this._processRun(run, fixes);
      });
    }
  }

  /**
   * Process a text run
   *
   * @param {Object} run - Run object
   * @param {Object} fixes - Fix counter
   * @private
   */
  _processRun(run, fixes) {
    if (!run['w:rPr']) {
      return;
    }

    const props = run['w:rPr'];

    // Fix font
    if (props['w:rFonts']) {
      this._normalizeFont(props['w:rFonts'], fixes);
    }

    // Fix color (convert theme colors to RGB)
    if (props['w:color']) {
      this._normalizeColor(props['w:color'], fixes);
    }

    // Fix bold (ensure consistent bold representation)
    if (props['w:b']) {
      this._normalizeBold(props['w:b'], fixes);
    }

    // Remove complex text effects that LibreOffice doesn't support
    const unsupportedEffects = ['w:shadow', 'w:outline', 'w:emboss', 'w:imprint', 'w14:glow', 'w14:shadow'];
    unsupportedEffects.forEach(effect => {
      if (props[effect]) {
        delete props[effect];
        fixes.stylesSimplified++;
      }
    });
  }

  /**
   * Normalize font specification
   *
   * @param {Object} fontSpec - Font specification object
   * @param {Object} fixes - Fix counter
   * @private
   */
  _normalizeFont(fontSpec, fixes) {
    const fontAttributes = ['@_w:ascii', '@_w:hAnsi', '@_w:cs', '@_w:eastAsia'];

    fontAttributes.forEach(attr => {
      if (fontSpec[attr]) {
        const originalFont = fontSpec[attr];
        const mappedFont = this.fontMap[originalFont] || originalFont;

        if (mappedFont !== originalFont) {
          fontSpec[attr] = mappedFont;
          fixes.fontsNormalized++;
        }
      }
    });

    // Remove theme font references (force explicit fonts)
    const themeAttributes = ['@_w:asciiTheme', '@_w:hAnsiTheme', '@_w:cstheme', '@_w:eastAsiaTheme'];
    themeAttributes.forEach(attr => {
      if (fontSpec[attr]) {
        delete fontSpec[attr];
      }
    });
  }

  /**
   * Normalize color (convert theme colors to RGB)
   *
   * @param {Object} colorSpec - Color specification
   * @param {Object} fixes - Fix counter
   * @private
   */
  _normalizeColor(colorSpec, fixes) {
    // If color uses theme reference, convert to explicit RGB
    if (colorSpec['@_w:themeColor']) {
      const themeColor = colorSpec['@_w:themeColor'];
      const rgbColor = this.themeColorMap[themeColor];

      if (rgbColor) {
        colorSpec['@_w:val'] = rgbColor;
        delete colorSpec['@_w:themeColor'];
        delete colorSpec['@_w:themeShade'];
        delete colorSpec['@_w:themeTint'];
        fixes.themeColorsConverted++;
      }
    }

    // Ensure 6-digit hex format
    if (colorSpec['@_w:val']) {
      let color = colorSpec['@_w:val'].toUpperCase();
      if (color === 'AUTO') {
        colorSpec['@_w:val'] = '000000'; // Black
      } else if (color.length === 3) {
        // Expand 3-digit hex to 6-digit
        colorSpec['@_w:val'] = color[0] + color[0] + color[1] + color[1] + color[2] + color[2];
      }
    }
  }

  /**
   * Normalize bold formatting
   *
   * @param {Object} boldSpec - Bold specification
   * @param {Object} fixes - Fix counter
   * @private
   */
  _normalizeBold(boldSpec, fixes) {
    // Ensure bold is explicitly enabled (not ambiguous)
    if (!boldSpec['@_w:val'] || boldSpec['@_w:val'] === 'true' || boldSpec['@_w:val'] === '1') {
      boldSpec['@_w:val'] = '1';
      fixes.boldFixed++;
    }
  }

  /**
   * Normalize paragraph properties
   *
   * @param {Object} pPr - Paragraph properties
   * @param {Object} fixes - Fix counter
   * @private
   */
  _normalizeParagraphProperties(pPr, fixes) {
    // Normalize spacing (convert relative to absolute)
    if (pPr['w:spacing']) {
      const spacing = pPr['w:spacing'];

      // Ensure numeric values
      if (spacing['@_w:before']) {
        spacing['@_w:before'] = String(parseInt(spacing['@_w:before']) || 0);
      }
      if (spacing['@_w:after']) {
        spacing['@_w:after'] = String(parseInt(spacing['@_w:after']) || 0);
      }
      if (spacing['@_w:line']) {
        spacing['@_w:line'] = String(parseInt(spacing['@_w:line']) || 240);
      }

      fixes.paragraphsAdjusted++;
    }

    // Remove widow/orphan control (LibreOffice handles differently)
    if (pPr['w:widowControl']) {
      delete pPr['w:widowControl'];
    }
  }

  /**
   * Process table
   *
   * @param {Object} table - Table object
   * @param {Object} fixes - Fix counter
   * @private
   */
  _processTable(table, fixes) {
    // Process table rows
    if (table['w:tr']) {
      const rows = Array.isArray(table['w:tr']) ? table['w:tr'] : [table['w:tr']];

      rows.forEach(row => {
        // Process cells
        if (row['w:tc']) {
          const cells = Array.isArray(row['w:tc']) ? row['w:tc'] : [row['w:tc']];

          cells.forEach(cell => {
            // Process cell paragraphs
            if (cell['w:p']) {
              const paragraphs = Array.isArray(cell['w:p']) ? cell['w:p'] : [cell['w:p']];
              paragraphs.forEach(p => this._processParagraph(p, fixes));
            }
          });
        }
      });
    }
  }

  /**
   * Process document styles
   *
   * @param {Object} stylesData - Parsed styles XML
   * @param {Object} fixes - Fix counter
   * @private
   */
  _processStyles(stylesData, fixes) {
    if (!stylesData['w:styles'] || !stylesData['w:styles']['w:style']) {
      return;
    }

    const styles = Array.isArray(stylesData['w:styles']['w:style'])
      ? stylesData['w:styles']['w:style']
      : [stylesData['w:styles']['w:style']];

    styles.forEach(style => {
      // Process style run properties
      if (style['w:rPr']) {
        if (style['w:rPr']['w:rFonts']) {
          this._normalizeFont(style['w:rPr']['w:rFonts'], fixes);
        }
        if (style['w:rPr']['w:color']) {
          this._normalizeColor(style['w:rPr']['w:color'], fixes);
        }
        if (style['w:rPr']['w:b']) {
          this._normalizeBold(style['w:rPr']['w:b'], fixes);
        }
      }

      // Process style paragraph properties
      if (style['w:pPr']) {
        this._normalizeParagraphProperties(style['w:pPr'], fixes);
      }

      fixes.stylesSimplified++;
    });
  }
}

module.exports = new DocxPreProcessor();

File: FileConversionApi/Services/DocxPreProcessor.cs
using Microsoft.Extensions.Logging;
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Wordprocessing;
using System.Text.RegularExpressions;
using System.Diagnostics;

namespace FileConversionApi.Services;

/// <summary>
/// DOCX Pre-Processor for improved LibreOffice compatibility
/// Normalizes formatting to reduce conversion issues
/// </summary>
public class DocxPreProcessor : IDocxPreProcessor
{
    private readonly ILogger<DocxPreProcessor> _logger;

    /// <summary>
    /// Font substitution map - replace problematic fonts with LibreOffice-friendly alternatives
    /// </summary>
    private readonly Dictionary<string, string> _fontMap = new()
    {
        // Office theme fonts -> standard fonts
        {"Calibri Light", "Calibri"},
        {"Segoe UI Light", "Segoe UI"},

        // Custom fonts -> safe alternatives
        {"Aptos", "Calibri"},
        {"Aptos Narrow", "Arial Narrow"},
        {"Arial Narrow", "Arial"},

        // Proprietary fonts -> open equivalents
        {"Times New Roman", "Liberation Serif"},
        {"Arial", "Liberation Sans"},
        {"Courier New", "Liberation Mono"},

        // Keep these safe fonts as-is (LibreOffice handles well)
        {"Calibri", "Calibri"},
        {"Verdana", "Verdana"},
        {"Georgia", "Georgia"},
        {"Tahoma", "Tahoma"},
    };

    /// <summary>
    /// Theme color mappings (Office theme colors to RGB)
    /// </summary>
    private readonly Dictionary<string, string> _themeColorMap = new()
    {
        {"accent1", "4472C4"}, // Blue
        {"accent2", "ED7D31"}, // Orange
        {"accent3", "A5A5A5"}, // Gray
        {"accent4", "FFC000"}, // Yellow
        {"accent5", "5B9BD5"}, // Light Blue
        {"accent6", "70AD47"}, // Green
        {"dark1", "000000"},   // Black
        {"dark2", "44546A"},   // Dark Gray
        {"light1", "FFFFFF"},  // White
        {"light2", "E7E6E6"},  // Light Gray
    };

    public DocxPreProcessor(ILogger<DocxPreProcessor> logger)
    {
        _logger = logger;
    }

    /// <inheritdoc/>
    public async Task<PreprocessingResult> ProcessAsync(string inputPath, string outputPath)
    {
        var fixes = new PreprocessingFixes();
        var stopwatch = Stopwatch.StartNew();

        try
        {
            _logger.LogInformation("Pre-processing DOCX for LibreOffice compatibility: {InputPath}", inputPath);

            // Copy input to output first
            await using (var inputStream = File.OpenRead(inputPath))
            await using (var outputStream = File.Create(outputPath))
            {
                await inputStream.CopyToAsync(outputStream);
            }

            // Process the document
            using var document = WordprocessingDocument.Open(outputPath, true);

            // Process main document
            if (document.MainDocumentPart?.Document != null)
            {
                ProcessDocument(document.MainDocumentPart.Document, fixes);
            }

            // Process styles
            if (document.MainDocumentPart?.StyleDefinitionsPart?.Styles != null)
            {
                ProcessStyles(document.MainDocumentPart.StyleDefinitionsPart.Styles, fixes);
            }

            // Remove page borders from all sections
            if (document.MainDocumentPart?.Document?.Body != null)
            {
                RemovePageBorders(document.MainDocumentPart.Document.Body, fixes);
            }

            document.Save();
            document.Dispose();

            stopwatch.Stop();

            _logger.LogInformation("DOCX pre-processing completed in {Time}ms", stopwatch.ElapsedMilliseconds);
            _logger.LogInformation("Fixes applied: Fonts={Fonts}, Colors={Colors}, Styles={Styles}, Bold={Bold}, PageBorders={PageBorders}",
                fixes.FontsNormalized, fixes.ThemeColorsConverted, fixes.StylesSimplified, fixes.BoldFixed, fixes.PageBordersRemoved);

            return new PreprocessingResult
            {
                Success = true,
                InputPath = inputPath,
                OutputPath = outputPath,
                Fixes = fixes,
                ProcessingTimeMs = stopwatch.ElapsedMilliseconds
            };
        }
        catch (Exception ex)
        {
            stopwatch.Stop();
            _logger.LogError(ex, "DOCX pre-processing failed");

            return new PreprocessingResult
            {
                Success = false,
                Error = $"Pre-processing failed: {ex.Message}",
                ProcessingTimeMs = stopwatch.ElapsedMilliseconds
            };
        }
    }

    /// <summary>
    /// Process main document content
    /// </summary>
    private void ProcessDocument(Document document, PreprocessingFixes fixes)
    {
        var body = document.Body;
        if (body == null) return;

        // Process paragraphs
        foreach (var paragraph in body.Descendants<Paragraph>())
        {
            ProcessParagraph(paragraph, fixes);
        }

        // Process tables
        foreach (var table in body.Descendants<Table>())
        {
            ProcessTable(table, fixes);
        }
    }

    /// <summary>
    /// Process paragraph properties and runs
    /// </summary>
    private void ProcessParagraph(Paragraph paragraph, PreprocessingFixes fixes)
    {
        // Process paragraph properties
        if (paragraph.ParagraphProperties != null)
        {
            NormalizeParagraphProperties(paragraph.ParagraphProperties, fixes);
        }

        // Process runs (text segments with formatting)
        foreach (var run in paragraph.Descendants<Run>())
        {
            ProcessRun(run, fixes);
        }
    }

    /// <summary>
    /// Process run properties (font, color, bold, etc.)
    /// </summary>
    private void ProcessRun(Run run, PreprocessingFixes fixes)
    {
        if (run.RunProperties == null) return;

        var runProps = run.RunProperties;

        // Normalize fonts
        if (runProps.RunFonts != null)
        {
            NormalizeFonts(runProps.RunFonts, fixes);
        }

        // Convert theme colors to RGB
        if (runProps.Color != null)
        {
            ConvertThemeColor(runProps.Color, fixes);
        }

        // Fix bold formatting
        if (runProps.Bold != null || runProps.BoldComplexScript != null)
        {
            FixBoldFormatting(runProps, fixes);
        }
    }

    /// <summary>
    /// Process table formatting
    /// </summary>
    private void ProcessTable(Table table, PreprocessingFixes fixes)
    {
        // Process table properties
        var tableProps = table.GetFirstChild<TableProperties>();
        if (tableProps != null)
        {
            // Could add table property normalization here if needed
        }

        // Process table rows and cells
        foreach (var row in table.Descendants<TableRow>())
        {
            foreach (var cell in row.Descendants<TableCell>())
            {
                foreach (var paragraph in cell.Descendants<Paragraph>())
                {
                    ProcessParagraph(paragraph, fixes);
                }
            }
        }
    }

    /// <summary>
    /// Process document styles
    /// </summary>
    private void ProcessStyles(Styles styles, PreprocessingFixes fixes)
    {
        foreach (var style in styles.Descendants<Style>())
        {
            if (style.StyleRunProperties != null)
            {
                // Process style-level formatting similar to run properties
                var styleRunProps = style.StyleRunProperties;

                if (styleRunProps.RunFonts != null)
                {
                    NormalizeFonts(styleRunProps.RunFonts, fixes);
                }

                if (styleRunProps.Color != null)
                {
                    ConvertThemeColor(styleRunProps.Color, fixes);
                }

                if (styleRunProps.Bold != null)
                {
                    FixBoldFormatting(styleRunProps, fixes);
                }
            }
        }
    }

    /// <summary>
    /// Normalize paragraph properties for better compatibility
    /// </summary>
    private void NormalizeParagraphProperties(ParagraphProperties paraProps, PreprocessingFixes fixes)
    {
        // Could add paragraph spacing normalization here if needed
        // For now, we focus on the most critical formatting issues
    }

    /// <summary>
    /// Normalize font references
    /// </summary>
    private void NormalizeFonts(RunFonts runFonts, PreprocessingFixes fixes)
    {
        var properties = runFonts.GetType().GetProperties();

        foreach (var prop in properties)
        {
            if (prop.PropertyType == typeof(string) && prop.Name.Contains("Theme"))
            {
                // Remove theme font references (force explicit fonts)
                prop.SetValue(runFonts, null);
                continue;
            }

            var fontName = prop.GetValue(runFonts) as string;
            if (!string.IsNullOrEmpty(fontName) && _fontMap.TryGetValue(fontName, out var replacement))
            {
                if (fontName != replacement)
                {
                    prop.SetValue(runFonts, replacement);
                    fixes.FontsNormalized++;
                }
            }
        }
    }

    /// <summary>
    /// Convert theme colors to explicit RGB values
    /// </summary>
    private void ConvertThemeColor(Color color, PreprocessingFixes fixes)
    {
        var themeColor = color.ThemeColor?.ToString();
        if (!string.IsNullOrEmpty(themeColor) && _themeColorMap.TryGetValue(themeColor, out var rgbColor))
        {
            color.Val = rgbColor;
            color.ThemeColor = null; // Remove theme reference
            fixes.ThemeColorsConverted++;
        }

        // Convert AUTO color to black
        if (color.Val?.Value == "AUTO")
        {
            color.Val = "000000";
        }
    }

    /// <summary>
    /// Fix bold formatting for consistency
    /// </summary>
    private void FixBoldFormatting(object runProps, PreprocessingFixes fixes)
    {
        var propsType = runProps.GetType();
        var boldProp = propsType.GetProperty("Bold");
        var boldCsProp = propsType.GetProperty("BoldComplexScript");

        // Fix regular bold
        if (boldProp != null)
        {
            var bold = boldProp.GetValue(runProps) as Bold;
            if (bold != null && bold.Val == null)
            {
                bold.Val = new DocumentFormat.OpenXml.OnOffValue(true);
                fixes.BoldFixed++;
            }
        }

        // Fix complex script bold
        if (boldCsProp != null)
        {
            var boldCs = boldCsProp.GetValue(runProps) as BoldComplexScript;
            if (boldCs != null && boldCs.Val == null)
            {
                boldCs.Val = new DocumentFormat.OpenXml.OnOffValue(true);
                fixes.BoldFixed++;
            }
        }
    }

    /// <summary>
    /// Remove page borders from all sections
    /// Fixes LibreOffice PDF conversion adding unwanted borders
    /// </summary>
    private void RemovePageBorders(Body body, PreprocessingFixes fixes)
    {
        // Find all section properties in the document
        var sectionProps = body.Descendants<SectionProperties>();

        foreach (var sectionProp in sectionProps)
        {
            // Find and remove PageBorders element if it exists
            var pageBorders = sectionProp.GetFirstChild<PageBorders>();
            if (pageBorders != null)
            {
                pageBorders.Remove();
                fixes.PageBordersRemoved++;
                _logger.LogDebug("Removed page border from section");
            }
        }
    }
}

/// <summary>
/// Preprocessing result
/// </summary>
public class PreprocessingResult
{
    public bool Success { get; set; }
    public string? InputPath { get; set; }
    public string? OutputPath { get; set; }
    public PreprocessingFixes? Fixes { get; set; }
    public long ProcessingTimeMs { get; set; }
    public string? Error { get; set; }
}

/// <summary>
/// Track what fixes were applied during preprocessing
/// </summary>
public class PreprocessingFixes
{
    public int FontsNormalized { get; set; }
    public int ThemeColorsConverted { get; set; }
    public int StylesSimplified { get; set; }
    public int ParagraphsAdjusted { get; set; }
    public int BoldFixed { get; set; }
    public int PageBordersRemoved { get; set; }
}

/// <summary>
/// DOCX preprocessor interface
/// </summary>
public interface IDocxPreProcessor
{
    Task<PreprocessingResult> ProcessAsync(string inputPath, string outputPath);
}
